src/interp.rs

```diff
@@ -55,97 +59,160 @@ pub struct TypedVal<'tcx> {
+                        block: BasicBlock) {
+        let basic_block_data = self.cur_frame().mir.basic_blocks().get(block).unwrap();
+        for statement in &basic_block_data.statements {
+            self.eval_statement(block, statement)
+        }
 
-/// The easy bit. `BasicBlock`s are a list of instructions which are guaranteed to
-/// always be executed sequentially. We just need to iterate over the statement
-/// list and execute each one in order.
-fn eval_basic_block(block: BasicBlock,
-                    vm: &mut Machine) {
-    let basic_block_data = vm.cur_frame().mir.basic_blocks().get(block).unwrap();
-    for statement in &basic_block_data.statements {
-        eval_statement(block, statement, vm)
+        let term = basic_block_data.terminator();
+        self.eval_terminator(term);
     }
 
-    let term = basic_block_data.terminator();
-    eval_terminator(term, vm);
-}
+    fn eval_statement(&mut self,
+                      block: BasicBlock,
+                      statement: &Statement<'tcx>) {
+        match statement.kind {
+            StatementKind::Assign(ref place, ref rvalue) => {
+                self.eval_assign(block, place, rvalue)
+            },
+            StatementKind::StorageLive(local) => {
+                // XXX: NotImplemented
+            },
+            StatementKind::StorageDead(local) => {
+                // XXX: NotImplemented
+            }
+            _ => unimplemented!()
+        }
+    }
 
-fn eval_statement<'a, 'tcx>(block: BasicBlock,
-                  statement: &Statement<'tcx>,
-                  vm: &mut Machine<'a, 'tcx>) {
-    match statement.kind {
-        StatementKind::Assign(ref place, ref rvalue) => {
-            eval_assign(block, place, rvalue, vm)
+    /// The RHS of an assignment is first evaluated to a primitive. This will be
+    /// either a pointer into memory or a primitive. It is then mapped to the
+    /// required place in memory.
+    fn eval_assign(&mut self,
+                   _block: BasicBlock,
+                   place: &Place<'tcx>,
+                   rvalue: &Rvalue<'tcx>) {
+        let dest = self.eval_place(place);
+        let ref dest_ty = place.ty(self.cur_frame().mir, self.tcx).to_ty(self.tcx);
+        match *rvalue {
+            Rvalue::Use(ref operand) => {
+                let val = self.eval_operand(operand).val;
+                let tyval = TypedVal {
+                    ty: dest_ty,
+                    val: val
+                };
+                self.store(tyval, dest);
+            },
+            Rvalue::Aggregate(ref kind, ref ops) => {
+                match **kind {
+                    AggregateKind::Tuple => {
+                        // XXX: Tuple layout needs some proper thought, and
+                        // should be upcoming in the next PR, so we handle the
+                        // empty tuple explicitly as this is the bottom type in
+                        // Rust.
+                        if !ops.len() == 0 {
```

Comment:
Any reason for the `! ... ==` instead of `!=`?

Reply:
No, this is stupid and I don't know why I did it